### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by Xymph** [hamstar/Wikimate#104]
Don't know if this is a common approach, but a discussion ticket seems a suitable way to talk about short-term plans for Wikimate updates.

I would like to wrap up #102 asap and move forward, as work on four new PRs is already done and tested. They are, in order:
1. issue #103
1. add table of contents to USAGE doc
1. change API calls from PHP format to JSON
1. issue #92

Regarding 3: PHP format is [deprecated ](https://www.mediawiki.org/wiki/API:Data_formats#Output)and only maintained for now due to [its popularity](https://www.mediawiki.org/wiki/API:Data_formats#Additional_notes). New users should stick [to JSON](https://www.mediawiki.org/wiki/API:Etiquette#Data_formats), and some API calls [like opensearch](https://www.mediawiki.org/wiki/API:Opensearch) don't even support PHP format.

The other reason to switch to JSON is that my implementation for 4. moves all API requests into a single request() method, which can then also contain the decoding step (and debug logging of request and response) in one central place, instead of the current multiple unserialize-s in methods like query(), edit() and upload(). And such centralizing can only be done if all requests use the same JSON format, as it is the only one supported by all API calls.

A format switch would ideally be verified with unit testing of those Wikimate methods, but this is an area I have little experience in, and it would be time-consuming to set this up. My own application scripts continue to work fine with the changes in 3. and 4. though.

The easy way (also avoiding potential merging conflicts) is to implement these PRs in the repo in the above order, so I'd prefer to just create and merge one branch at a time (although 1-2 can be done in parallel with 3-4). But that brings me back to the (slow) progress of #102, and @waldyrious 's availability here. I certainly don't want to you as a punching bag my friend, and my afterthought commits in #102 demonstrate that reviews/feedback on my stuff is no luxury :wink: , but I would prefer not to stay stalled for long(er) times either.

Please share any thoughts on the above.

As for the longer term, I intend to look into existing issues again, and possibly explore coverage of more API features, but have no definite ideas/plans yet.

**Comment by Xymph** [comment]
I just ran phpdoc on Wikimate.php for the first time, and it pointed out some missing summaries, mostly for properties. So another planned PR is to add the absent info.

**Comment by waldyrious** [comment]
> Don't know if this is a common approach, but a discussion ticket seems a suitable way to talk about short-term plans for Wikimate updates.

Yeah, I agree that roadmap issues like this sound like a great way to plan out work.

As for the changes themselves and their proposed implementation schedule, they sound sensible. I don't have any objections to what you've laid out.

Finally, regarding the feedback cycle and development pace of the repo: believe I should be able to respond to issues/PRs in this repo in a few days (less than a week for sure) in the foreseeable future. I don't want to be a bottleneck on the repo, though, so if you see that your own availability or drive can't sustain my pace of feedback, I'd rather you go ahead and make changes by yourself, than have the project lose steam out of trying to stick to our proscribed review process (especially since my reviews will be superficial at best). That should be a guideline, not a blind rule.

**Comment by Xymph** [comment]
Thanks for your initial comments on my planned updates. And while this looks like a flurry of activity now, once they're done we'll be able to catch our breaths again :smile: and any following PRs will likely occur at the much slower pace from before.

**Comment by Xymph** [comment]
Well, 0.13.0 took a lot more polishing work than I expected last week, but some of that was residue of older work, and the release came out better in the end. Thanks for all your help, @waldyrious. :+1: 

FYI, I've updated "our" entry on [MediaWiki's library list](https://www.mediawiki.org/wiki/API:Client_code/All#PHP) now, and earlier already updated Wikipedia's [bot table](https://en.wikipedia.org/wiki/Help:Creating_a_bot#PHP) (actually [here](https://en.wikipedia.org/wiki/Wikipedia:PHP_bot_framework_table)). The only non-English wiki I found with such a table is the Turkish [one](https://tr.wikipedia.org/wiki/Yard%C4%B1m:Bot_olu%C5%9Fturma#PHP), also [updated](https://tr.wikipedia.org/wiki/Vikipedi:PHP_bot_%C3%A7er%C3%A7eve_tablosu).

**Comment by waldyrious** [comment]
Yay, it's finally out! I agree the extra time we spent on it paid off. 👍

Great call with updating the wiki pages as well.

Looking forward to the next changes!

**Comment by Xymph** [comment]
Two further ideas:
- support the [exclusion standard](https://en.wikipedia.org/wiki/Template:Bots) where we currently [score a No](https://en.wikipedia.org/wiki/Help:Creating_a_bot#PHP)
- post the PHPDoc generated documentation tree somewhere publicly

**Comment by waldyrious** [comment]
> post the PHPDoc (or maybe ApiGen) generated documentation tree somewhere publicly

I like this! Perhaps GitHub Pages would be a good place to host these docs.

**Comment by Xymph** [comment]
> > post the PHPDoc (or maybe ApiGen) generated documentation tree somewhere publicly
> 
> I like this! Perhaps GitHub Pages would be a good place to host these docs.

Yes, that was my first inclination too, but being unfamiliar with it, I'll need to read up on how to place/update content there.

**Comment by waldyrious** [comment]
If we can use GitHub Actions to automatically generate the HTML documentation and add them to a `docs/` folder, then serving them from GitHub Pages should be as simple as this:

![Screenshot 2021-07-06 at 23 04 52](https://user-images.githubusercontent.com/478237/124672513-0afe1980-deaf-11eb-9217-8a226c2cb951.png)

We would need access to the repository settings, though. Currently only @hamstar has that level of access. I'll email him to see if he's up for it.

**Comment by Xymph** [comment]
Better not to clutter this high-level topic with the details of one task, so I created #110. Maybe you could move your last post into that thread?

**Comment by waldyrious** [comment]
#111 got me thinking — normally, such a change woul require a major version bump, since we're changing the tool's interface. However, we're still using a 0.x "development" version scheme, so that wouldn't apply in this case. Still, perhaps this change would be a good moment to mark 1.0 (unless you're planning further backward-incompatible changes). WDYT? Should we make the next version 1.0?

**Comment by Xymph** [comment]
> #111 got me thinking — normally, such a change would require a major version bump, since we're changing the tool's interface. However, we're still using a 0.x "development" version scheme, so that wouldn't apply in this case. Still, perhaps this change would be a good moment to mark 1.0 (unless you're planning further backward-incompatible changes). WDYT? Should we make the next version 1.0?

If this were an API breaking change I'd have pointed it out and my client scripts wouldn't have continued to run unmodified. The Wikimate methods all continue to return associated array structures: the only thing changed by the PR is how those were requested from the API and decoded. `PHP+unserialize == JSON+json_decode(associated=true)`.

So no, the next release can simply be 0.14.0. When the 1.0.0 release would be appropriate I don't know yet - perhaps when Wikimate covers most of the API, but I doubt it'll ever get expanded that far.

**Comment by waldyrious** [comment]
Ah, I see. Yeah, I did get that impression from what you wrote in the PR description, but while reading the code diff, I didn't make the connection that "Unserialized php" = "Decoded JSON" and my brain snapped back into the assumption the output was the same content with a sightly different representation :)

As for 1.0, I get what you mean, but OTOH Semantic Versioning does [recommend](https://semver.org/#how-do-i-know-when-to-release-100)

> If your software is being used in production, it should probably already be 1.0.0. If you have a stable API on which users have come to depend, you should be 1.0.0. If you’re worrying a lot about backwards compatibility, you should probably already be 1.0.0.

At least two of those are true, for what that's worth. Releasing the first major version would also allow us to make our semantic versions more, well, semantic. But I definitely get the appeal of leaving 1.0 to when the coverage of the API is closer to complete. I'm OK with whatever you prefer.

**Comment by Xymph** [comment]
> As for 1.0, I get what you mean, but OTOH Semantic Versioning does [recommend](https://semver.org/#how-do-i-know-when-to-release-100)

Thanks for digging that up. Yes, I agree we're sufficiently out of development to go for 1.0. But it seems fitting to have the public API documentation site online when we do that. Also, I noticed that a [Logout call](https://www.mediawiki.org/wiki/Special:MyLanguage/API:Logout) exists which is easy to implement, so it would be a cleaner 1.0 if it supports both login and logout. However little used the latter may be, since one is effectively also logged out once the client script completes. I never missed it, anyway. :smile:

**Comment by panda87** [comment]
Where is @glatomer to approve it?
He is the maintainer, isn't he?

**Comment by Xymph** [comment]
> Where is @glatomer to approve it?
> He is the maintainer, isn't he?

galtomer (not gla...) set up the package on Packagist but it turned out I was also able [to update it](https://github.com/hamstar/Wikimate/pull/102#issuecomment-874649278) after creating an account. So we're good.

**Comment by panda87** [comment]
Cool. I was afraid we will need to wait for him... he is not too responsive for some reason

**Comment by Xymph** [comment]
Re. v1.0.0, a major version bump is also required when making API breaking changes. If error handling/exceptions are changed (#113) and/or debug logging is changed into levels (already suggested in #101), then those should be completed before the bump to 1.0, as it would be awkward to bump to 2.0.0 soon after.

**Comment by Xymph** [comment]
Didn't want to post this sooner so as not to distract @waldyrious from more important matters (:smile:) but here are my plans for the next few releases (all code for 0.15.0 already prepared):

- wrap up v0.14.0 (it's big enough and includes requested fixes)
---
- fix user agent format (is missing a slash)
- address #115
- add logout method
- add file revert method
- wrap up v0.15.0
---
- remove obsolete method debugCurlConfig
- fix #116
- use consistent error return / exception pattern and resolve #113; describe in Usage.md
- address #110
- wrap up v1.0.0
---
- support exclusion standard
- file usage method?
- page revisions per #82?
- more classes for categories, searching, users?

Any feedback?

**Comment by waldyrious** [comment]
Sounds great to me! Looking forward to the documentation part, where I suspect I might be able to help a bit more. The authentication feature would be a pretty nice addition too.

**Comment by Xymph** [comment]
Since it would otherwise fly under the radar, I better point out that yesterday I added five new entries to the post [v1.0 roadmap above](https://github.com/hamstar/Wikimate/issues/104#issuecomment-903944692):

- implement supported parameter mechanism based on [paraminfo](https://www.mediawiki.org/wiki/Special:MyLanguage/API:Paraminfo) API request
- support [deletetalk parameter](https://www.mediawiki.org/wiki/API:Delete#Parameter_history) in delete page & file methods
- support [undelete](https://www.mediawiki.org/wiki/Special:MyLanguage/API:Undelete) API request
- support [siteinfo](https://www.mediawiki.org/wiki/Special:MyLanguage/API:Siteinfo) API request?
- apply [JSON format version 2](https://www.mediawiki.org/wiki/API:JSON_version_2)?

With probably more to follow.
